### PR TITLE
Initial transformer masking logic

### PR DESCRIPTION
**Pull request by lbaierreinio**
Disregard changes to the config files/cs-cluster scripts. More detailed explanations in comments below

### CODE REVIEW

File: src/datasets/synthetic/generate_dataset.py [hunk 1]
```diff
@@ -40,13 +40,8 @@ def main():
...
 
-    transformer_tokenizer = AutoTokenizer.from_pretrained(
-        config.tokenizer, model_max_length=config.sequence_length+config.cls_tokens)
-
-    mingru_tokenizer = AutoTokenizer.from_pretrained(
-        config.tokenizer, model_max_length=config.sequence_length)
-
-    mingru_tokenizer.padding_side = "left" # MinGRU uses pre-padding
+    tokenizer = AutoTokenizer.from_pretrained(config.tokenizer)
+    tokenizer.padding_side = "left"
```
Comment (lbaierreinio) [inline src/datasets/synthetic/generate_dataset.py:44]: No need for a separate tokenizer for minGRU and Transformer, as they both use pre padding now (necessary b/c chunked transformer uses RNN layer at the end)

File: src/datasets/synthetic/utility.py [hunk 2]
```diff
@@ -47,7 +48,7 @@ def generate_dataset8(*, seq_len, num_examples, grammars, alpha, beta, k_split,
         order = orders[label]
 
         # Draw sequence length from beta distribution
-        cur_seq_len =  max(32, int(np.random.beta(alpha, beta) * seq_len))
+        cur_seq_len = min_seq_len + int(np.random.beta(alpha, beta) * (max_seq_len - min_seq_len))
```
Comment (lbaierreinio) [inline src/datasets/synthetic/utility.py:51]: Provide a configurable lower bound for sequence length

File: src/train/utility.py [hunk 3]
```diff
@@ -28,43 +29,48 @@ def evaluate(model, dataloader, loss_fn, evaluation_type='Validation'):
...
-            total_loss += loss.item()
+            total_loss += loss.detach().item()
             loss.backward()
-            optimizer.step()
+
+            # Gradient accumulation every i batches (ensure number of batches divisible by i)
+            if i % accumulate_every_i == 0:
+                optimizer.step()
+                optimizer.zero_grad()
+
```
Comment (lbaierreinio) [inline src/train/utility.py:56]: Allow for gradient accumulation as our batch size may be constrained when training on longer sequence lengths

File: src/train/utility.py [hunk 4]
```diff
@@ -28,43 +29,48 @@ def evaluate(model, dataloader, loss_fn, evaluation_type='Validation'):
...
-        if epoch % 5 == 0:
+        if (epoch + 1) % check_every_i == 0:
             print(f"------------ EPOCH {epoch} ------------ \n")
             print(f"Training Loss: {round(total_loss, 4)}")
             total_loss, accuracy = evaluate(
                 model, val_dataloader, loss_fn, 'Validation')
             if early_stopping_threshold is not None and accuracy >= early_stopping_threshold:
                 print(f"Early stopping at epoch {epoch}")
-                return total_loss, accuracy, steps, epoch, (total_time / steps)
+                return total_loss, accuracy, steps, epoch, (total_time / num_epochs)
```
Comment (lbaierreinio) [inline src/train/utility.py:71]: Avg total time per epoch rather than total time per step

File: src/train/utility.py [hunk 5]
```diff
@@ -11,7 +11,8 @@ def evaluate(model, dataloader, loss_fn, evaluation_type='Validation'):
         for batch in dataloader:
             input = batch['input_ids'].to(device)
             labels = batch['labels'].to(device)
-            output = model(input)
+            mask = ~batch['attention_mask'].to(device).bool()
```
Comment (lbaierreinio) [inline src/train/utility.py:14]: Negate mask as per our discussion

File: src/datasets/synthetic/MinGRUSyntheticDataset.py [hunk 6]
```diff
@@ -3,12 +3,11 @@
...
 
 class MinGRUSyntheticDataset(Dataset):
-    def __init__(self, sequences, labels, tokenizer, max_length):
+    def __init__(self, sequences, labels, tokenizer):
         self.encodings = tokenizer(
             [' '.join(seq) for seq in sequences],
-            truncation=False,
+            padding=True,
             add_special_tokens=True,
-            max_length=max_length,
```
Comment (lbaierreinio) [inline src/datasets/synthetic/MinGRUSyntheticDataset.py:11]: No need to pad to max length for minGRU

File: src/datasets/synthetic/TransformerSyntheticDataset.py [hunk 7]
```diff
@@ -3,17 +3,13 @@
 
 
 class TransformerSyntheticDataset(Dataset):
-    def __init__(self, sequences, labels, tokenizer, max_length=2048, chunk_size=512):
-        for sequence in sequences:
-            for i in range(0, len(sequence), chunk_size):
-                sequence.insert(i, tokenizer.cls_token)
```
Comment (lbaierreinio) [inline src/datasets/synthetic/TransformerSyntheticDataset.py:9]: No longer need [CLS] tokens since minGRU 'attends' over all outputs of the Transformer + MinGRU model

File: src/datasets/synthetic/TransformerSyntheticDataset.py [hunk 8]
```diff
@@ -3,17 +3,13 @@
...
+    def __init__(self, sequences, labels, tokenizer, max_length=2048):
 
         self.encodings = tokenizer(
             [' '.join(seq) for seq in sequences],
-            truncation=False,
-            add_special_tokens=False,
             padding='max_length',
-            max_length=max_length,
+            add_special_tokens=False,
+            max_length = max_length,
```
Comment (lbaierreinio) [inline src/datasets/synthetic/TransformerSyntheticDataset.py:12]: The Transformer dataset pads up to max length to ensure that the chunks are evenly divisible

File: src/experiments/train_model.py [hunk 9]
```diff
@@ -104,17 +103,11 @@ def main():
...
         ).to(device)
 
     num_parameters = sum(p.numel() for p in model.parameters())
 
-    print(f"Model: {config.name}")
-    print(f"Number of Parameters: {num_parameters}")
-    print(f"Bidirectional: {config.bidirectional}")
-    print(f"Number of Layers: {config.num_layers}")
-    print(f"Embedding Dimension: {config.embedding_dim}")
-
```
Comment (lbaierreinio) [inline src/experiments/train_model.py:117]: Going to define a systematic way to log results in next PR

File: src/layers/transformer/LongTransformerEncoder.py [hunk 10]
```diff
@@ -19,34 +20,29 @@ def __init__(self, vocab_size, num_heads, num_layers, num_hiddens=512, ffn_num_h
             TransformerEncoderBlock(num_heads, num_hiddens, ffn_num_hiddens, dropout, bias) for _ in range(num_layers)
         ])
 
-        # Aggregate result of each chunk
-        self.rnn_out = nn.GRU(num_hiddens, num_hiddens, num_layers=1, batch_first=True)
+        # MinGRU for output
+        self.out = BiMinGRU(num_hiddens, num_hiddens)
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/layers/transformer/LongTransformerEncoder.py:24]: Detailed this a bit in the document I shared earlier today: https://docs.google.com/document/d/15fit6uYm3XCPMRXpZJabhBTH6wFjj6H8_u4z_lqvN9Q/edit?tab=t.0#heading=h.f2ch0vowiyoc Open to suggestions & thoughts on it

File: src/train/utility.py [hunk 11]
```diff
@@ -28,43 +29,48 @@ def evaluate(model, dataloader, loss_fn, evaluation_type='Validation'):
...
             mask = ~batch['attention_mask'].to(device).bool()
-
-            start = time.time()  # TODO: Verify this is a good way to measure time b/c batch size might not necessarily always be the same(?)
-            optimizer.zero_grad()
+            t0 = time.time()
+            # Backward & forward pass
             output = model(input, mask=mask)
             loss = loss_fn(output, labels)
-            total_loss += loss.item()
+            total_loss += loss.detach().item()
```
[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/train/utility.py:49]: I did add a comment indicating the number of batches should be divisible by accumulate_every_i, but I can also add an assertion to verify this. I think in practice I probably won't have to change accumulate_every_i from its default value, which is two.

 Are you sure about that first point? I implemented my solution based on what I read online about gradient accumulation (e.g. https://wandb.ai/wandb_fc/tips/reports/How-To-Implement-Gradient-Accumulation-in-PyTorch--VmlldzoyMjMwOTk5)

File: src/layers/transformer/LongTransformerEncoder.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dlrudwo1269) [reply src/layers/transformer/LongTransformerEncoder.py:24]: Ah okay makes sense. Smart!

Personally I would like to see just the chunked transformer (completely decoupled from minGRU) for comparison's sake, but agree this would likely work better

[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/layers/transformer/LongTransformerEncoder.py:24]: Agreed. I did add a note to that effect to verify that it doesn't work as well. I'll put priority on these models first, but then circle back to that at the end

File: src/train/utility.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dlrudwo1269) [reply src/train/utility.py]: Have a look at https://stackoverflow.com/a/65913698

[the remark this answers is not in the document]
Comment (dlrudwo1269) [reply src/train/utility.py]: I guess even in that article you linked, if you scroll down to the next section it divides loss by NUM_ACCUMULATION_STEPS

[the remark this answers is not in the document]
Comment (lbaierreinio) [reply src/train/utility.py]: Lmao my bad
